### PR TITLE
Split files for PSR-4 compliance

### LINKED ISSUE #25: Use PSR-4 autoloading
Use PSR-4 autoloading, explode exceptions to single-class files

Original issue in https://github.com/grogy/PHP-Parallel-Lint/issues/2 thanks to @szepeviktor

### LINKED ISSUE #26: Rename Jakub's namespace
Original in https://github.com/grogy/PHP-Parallel-Lint/issues/3 thanks to @szepeviktor

### PR DESCRIPTION
**Pull request by roelofr**
There were many classes combined into single files, which isn't recommended and makes maintaining hard.

For example; Did you know that `\JakubOnderka\PhpParallelLint\ArrayIterator` was actually located in `src/Settings.php`. Not where you'd expect it to be.

This PR moves all these classes to their own files, and configures Composer to use a PSR-4 autoloader instead of a classmap.

Closes #25
Closes #26

### CODE REVIEW

**Review by jrfnl — COMMENTED** [review]
While I agree that this is a step in the right direction, I'd like to suggest taking things a step (or two) further if @grogy agrees.

Suggestions:
1. Implement #26 (rename the namespace) at the same time.
2. Use more subdirectories to place classes in logical groups.
    For example (more groupings may be possible, but I haven't looked at everything):
    - `Output` for the `Output` interface and all classes extending it.
    - `Exception[s]` for all `Exceptions`

Yes, that would be a BC-break and necessitate a major release, but that is needed anyhow when addressing #26, so we may as well try to set it up in a way which will make maintenance more intuitive for the future.

**Comment by roelofr** [comment]
Yeah, when I was refactoring the code I noticed that my `src/` folder was growing at a rapid pace, with `Exceptions` not being namespaced.

I dabbled about doing that too, but being a first-time contributor to this project, that seemed a bridge too far to cross on the first PR.

**Comment by jrfnl** [comment]
@grogy Could you please pitch in to say whether you agree with the suggestions above ?

Would be great if we could get this PR moving again to eventually get it merged.

**Comment by roelofr** [comment]
Okay, the codebase is split up into some folders, and I ran phpstan to make sure all links were correct.

Unti tests still fail locally (something about `JsonSerializable` not being found, while `ext-json` is active).

File: src/Exceptions/CallbackNotImplementException.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?php
+namespace PhpParallelLint\PhpParallelLint\Exceptions;
+
+class CallbackNotImplementException extends ParallelLintException
```
Comment (glensc) [inline src/Exceptions/CallbackNotImplementException.php:4]:
```suggestion
class CallbackNotImplementedException extends ParallelLintException
```

Comment (roelofr) [reply src/Exceptions/CallbackNotImplementException.php]: Thanks, fixed it in the rebase onto the updated master.

**Comment by grogy** [comment]
I readed changes commit to commit and it makes sense. I see a problem that I missed that PR and codebase is far away.

For resolve and merge I can rebase and solve it locally and force push it, or you can rebase it onto current master?

**Comment by roelofr** [comment]
I've rebased the commits, the only big change seemed to have been the GitLabOutput

**Comment by roelofr** [comment]
I noticed the tests were failing on a missing polyfill, so I've added that back in.

**Comment by jrfnl** [comment]
> I noticed the tests were failing on a missing polyfill, so I've added that back in.

Please load that file via a `require_once` in the bootstrap instead of letting Composer autoload it. See the dicussion about this here: https://github.com/php-parallel-lint/PHP-Parallel-Lint/pull/51#discussion_r593924479

**Comment by roelofr** [comment]
@jrfnl I tried removing it, but that doesn't work in combination with PHP 5.3, so I re-added it as `autoload-dev`, which should prevent it from loading in projects requiring this package.

**Comment by jrfnl** [comment]
> I tried removing it, but that doesn't work in combination with PHP 5.3, so I re-added it as autoload-dev, which should prevent it from loading in projects requiring this package.

Using `autoload-dev` feels wrong as the polyfill is needed for non-dev when the JSON extension is not available. Hmm.... I don't have to time to look into this further at the moment, but I do think we need to find another solution.

**Comment by roelofr** [comment]
I've removed the commit for now, but as said, it won't work.

The assumption made in the comment mentioned by @jrfnl is, in my opinio, out of scope.

> This would prevent potential issues with other dev tools which may use an `interface_exists()` check on `JsonSerializable` which would now suddenly start passing for them due to the file being automatically loaded (even when this tool is not being run).

If devtools use the existence on `JsonSerializable` to check if json is installed, they won't work on php 5.3 to begin with.

Wrapping the polyfill in a `version_compare(PHP_VERSION, '5.3', '<=')` should therefore suffice, and we can still include the polyfill.

**Comment by roelofr** [comment]
@jrfnl Did you manage to think about how we're resolving the polyfill issue? I think including it in `composer.json` is okay, as it's what we're doing in our current latest version... :thinking:

**Comment by jrfnl** [comment]
I've taken the liberty to rebase this PR and add an extra commit changing the namespace prefix to `PHP_Parallel_Lint\PhpParallelLint` to be in line with the other repos in this organisation.

Will have a look at the PHP 5.3 failure now.

**Comment by jrfnl** [comment]
(sorry for the long wait, but 🤞🏻 we'll get this merged now soon)

**Comment by jrfnl** [comment]
I've added an extra commit to sort out the test run on PHP 5.3.

As the polyfilled class will no longer be loaded via the Composer autoload file (as it doesn't have the expected namespace), the file needs to be `require`d in the tests.

When the tests got refactored to PHPUnit, these `require`s will move to a test bootstrap file instead.

**Review by jrfnl — APPROVED** [review]
@grogy This is ready for merge once PR #92 has been merged and version `1.3.2` has been released.

Might be an idea to (temporarily) create a `develop` branch and merge this PR to that branch so the changes for the `2.0.0 version are separated out and won't block a potential `1.3.3` release.

**Comment by glensc** [comment]
`PHP_Parallel_Lint` as namespace is terrible idea. It likely conflicts with psr-0 that composer has for legacy pear packages. or create some other unknown issues. avoid that. just remove underscores.

**Comment by jrfnl** [comment]
> `PHP_Parallel_Lint` as namespace is terrible idea. It likely conflicts with psr-0 that composer has for legacy pear packages. or create some other unknown issues. avoid that. just remove underscores.

I hear you, but a) conflicts will only happen if someone would be using the same namespace and b) the `PHP_Parallel_Lint` prefix has already been implemented in Console Color and Console Highlighter - using a different prefix here would be terribly inconsistent.

**Comment by roelofr** [comment]
Personally, it stings that the PR I made is now made so ugly with this new namespace change (which I think is _really_ ugly, and a step backwards from no namespaces), but it's been nearly a year and I can't really care anymore.

Good luck with this PR and the project, and thanks for looking into this after a year

**Comment by jrfnl** [comment]
> Personally, it stings that the PR I made is now made so ugly with this new namespace change (which I think is _really_ ugly, and a step backwards from no namespaces), but it's been nearly a year and I can't really care anymore.
> 
> Good luck with this PR and the project, and thanks for looking into this after a year

@roelofr I'm sorry you feel this way and I empathize with your frustration about how long the PR has been open. Has happened to me too often as well. The joys of open source....

**Comment by jrfnl** [comment]
Rebased without changes after the merge of #92 to get past the imaginary merge conflict.

**Pull request merged**